cong-framework/python/core/function.py

```
@@ -0,0 +1,150 @@
+        if best_generic_impl is not None:
+            return best_generic_impl.apply_value(*evaluated_args)
+
+        # If we get here, neither specific nor generic implementations worked
+        return Undefined(f"No applicable generic implementation with defined and valid costs found for {self.specification.name}.")
+
+def FunctionFromSpec(*args, **kwargs):
+    return Exp(_FunctionFromSpec(*args, **kwargs))
+
+class _GenericFunction(Base):
+
+    def __init__(self, impl, time_complexity=Undefined("Operation time cost not defined"), space_complexity=Undefined("Operation space cost not defined")):
```

Ruff (0.8.2)

Do not perform function call `Undefined` in argument defaults; instead, perform the call within the function, or read the default from a module-level singleton variable

(B008)